internal/locate/region_cache.go

```diff
@@ -2207,15 +2212,78 @@ func (c *RegionCache) batchScanRegions(bo *retry.Backoffer, keyRanges []pd.KeyRa
 
 		metrics.RegionCacheCounterWithBatchScanRegionsOK.Inc()
 		if len(regionsInfo) == 0 {
-			return nil, errors.Errorf(
+			backoffErr = errors.Errorf(
 				"PD returned no region, range num: %d, limit: %d",
 				len(keyRanges), limit,
 			)
+			continue
+		}
+		if regionsHaveGapInRanges(keyRanges, regionsInfo, limit) {
+			backoffErr = errors.Errorf(
+				"PD returned regions have gaps, range num: %d, limit: %d",
+				len(keyRanges), limit,
+			)
+			continue
 		}
 		return c.handleRegionInfos(bo, regionsInfo, opt.needRegionHasLeaderPeer)
 	}
 }
 
+// regionsHaveGapInRanges checks if the loaded regions can fully cover the key ranges.
+// If there are any gaps between the regions, it returns true, then the requests might be retried.
+// TODO: remove this function after PD client supports gap detection and handling it.
+func regionsHaveGapInRanges(ranges []pd.KeyRange, regionsInfo []*pd.Region, limit int) bool {
+	if len(ranges) == 0 {
+		return false
+	}
+	if len(regionsInfo) == 0 {
+		return true
+	}
+	checkIdx := 0                  // checked index of ranges
+	checkKey := ranges[0].StartKey // checked key of ranges
+	for _, r := range regionsInfo {
+		if r.Meta == nil {
+			return true
+		}
+		if len(r.Meta.StartKey) != 0 && bytes.Compare(r.Meta.StartKey, checkKey) > 0 {
+			// there is a gap between returned region's start_key and current check key
+			return true
+		}
+		if len(r.Meta.EndKey) == 0 {
+			// the current region contains all the rest ranges.
+			return false
+		}
+		checkKey = r.Meta.EndKey
+		for len(ranges[checkIdx].EndKey) > 0 && bytes.Compare(checkKey, ranges[checkIdx].EndKey) >= 0 {
```

Comment:
Not very familiar with the usage of it. Is the ranges here guaranteed to be already sorted by their start key?

Reply:
Yes, the input ranges are sorted and cannot be overlapped.